cmd/helm/repo_list.go

```diff
@@ -45,22 +56,71 @@ func newRepoListCmd(out io.Writer) *cobra.Command {
 		},
 	}
 
+	f := cmd.Flags()
+	f.StringVarP(&list.output, "output", "o", "", "output the specified format (json or yaml)")
 	return cmd
 }
 
 func (a *repoListCmd) run() error {
-	f, err := repo.LoadRepositoriesFile(a.home.RepositoryFile())
+	repoFile, err := repo.LoadRepositoriesFile(a.home.RepositoryFile())
 	if err != nil {
 		return err
 	}
-	if len(f.Repositories) == 0 {
-		return errors.New("no repositories to show")
-	}
-	table := uitable.New()
-	table.AddRow("NAME", "URL")
-	for _, re := range f.Repositories {
-		table.AddRow(re.Name, re.URL)
+
+	output, err := formatRepoListResult(a.output, repoFile)
+
+	if err != nil {
+		return err
 	}
-	fmt.Fprintln(a.out, table)
+	fmt.Fprintln(a.out, output)
+
 	return nil
 }
+
+func formatRepoListResult(format string, repoFile *repo.RepoFile) (string, error) {
+	var output string
+	var err error
+
+	if len(repoFile.Repositories) == 0 {
+		err = fmt.Errorf("no repositories to show")
+		return output, err
+	}
+
+	switch format {
+	case "":
+		table := uitable.New()
+		table.AddRow("NAME", "URL")
+		for _, re := range repoFile.Repositories {
+			table.AddRow(re.Name, re.URL)
+		}
+		output = table.String()
+
+	case "json":
+		var repolist repositories
+		for _, re := range repoFile.Repositories {
+			repolist.Repositories = append(repolist.Repositories, &repositoryElement{Name: re.Name, URL: re.URL})
+		}
+
+		o, e := json.Marshal(repolist)
```

Comment:
Much of this code in all of the printers is pretty repetitive and could be refactored out to be a bit more generic. I think to make this a little bit more maintainable long term, we should probably invest in doing that. For example, you could combine the for loops putting together the lists into a single case statement and then have a `PrintFormatted(obj interface{}, type string)` that can then marshal the data appropriately. That way we don't have duplicate error handling and marshaling everywhere. Another possibility would be to have all of these element/container types implement some sort of interface that has methods for printing a table, json, or yaml. In any case, there are multiple ways to solve this problem and I'd be open to any of them. Let me know if I wasn't clear or if you have any questions

Reply:
By the way, feel free to ping me on Slack if you have any questions or would like to pair